hswidget/cof_editor_widget.go

```diff
@@ -0,0 +1,529 @@
+			giu.Combo("##"+p.id+"AddLayerTransparent", trueFalse[state.COFEditorState.newLayerTransparent],
+				trueFalse, &state.COFEditorState.newLayerTransparent).Size(trueFalseListW),
+		),
+		giu.Line(
+			giu.Label("Draw effect: "),
+			giu.Combo("##"+p.id+"AddLayerDrawEffect", drawEffectList[state.COFEditorState.newLayerDrawEffect],
+				drawEffectList, &state.COFEditorState.newLayerDrawEffect).Size(bigListW),
+		),
+		giu.Line(
+			giu.Label("Weapon class: "),
+			giu.Combo("##"+p.id+"AddLayerWeaponClass", weaponClassList[state.COFEditorState.newLayerWeaponClass],
+				weaponClassList, &state.COFEditorState.newLayerWeaponClass).Size(bigListW),
+		),
+		giu.Separator(),
+		giu.Line(
+			giu.Button("Save##AddLayer").Size(saveCancelButtonW, saveCancelButtonH).OnClick(func() {
+				newCofLayer := &d2cof.CofLayer{
+					Type:        d2enum.CompositeType(state.COFEditorState.newLayerType),
+					Shadow:      byte(state.COFEditorState.newLayerSelectable),
+					Selectable:  (state.COFEditorState.newLayerSelectable == 1),
+					Transparent: (state.COFEditorState.newLayerTransparent == 1),
+					DrawEffect:  d2enum.DrawEffect(state.COFEditorState.newLayerDrawEffect),
+					WeaponClass: d2enum.WeaponClass(state.COFEditorState.newLayerWeaponClass),
+				}
+
+				p.cof.CofLayers = append(p.cof.CofLayers, *newCofLayer)
+
+				p.cof.NumberOfLayers++
+
+				for i := range p.cof.Priority {
+					for j := range p.cof.Priority[i] {
+						p.cof.Priority[i][j] = append(p.cof.Priority[i][j], newCofLayer.Type)
+					}
+				}
+
+				state.state = cofEditorStateViewer
+			}),
+			giu.Button("Cancel##AddLayer").Size(saveCancelButtonW, saveCancelButtonH).OnClick(func() {
+				state.state = cofEditorStateViewer
+			}),
+		),
+	}
+}
+
+func (p *COFWidget) deleteCurrentLayer(index int32) {
+	p.cof.NumberOfLayers--
+
+	newLayers := make([]d2cof.CofLayer, 0)
+
+	for n, i := range p.cof.CofLayers {
+		if int32(n) != index {
+			newLayers = append(newLayers, i)
+		}
+	}
+
+	p.cof.CofLayers = newLayers
+}
+
+func (p *COFWidget) duplicateDirection() {
+	stateID := fmt.Sprintf("COFWidget_%s", p.id)
+	s := giu.Context.GetState(stateID)
+
+	state := s.(*COFState)
+
+	idx := state.COFViewerState.directionIndex
+
+	p.cof.NumberOfDirections++
+
+	p.cof.Priority = append(p.cof.Priority, p.cof.Priority[idx])
+
+	// nolint:gomnd // directionIndex starts from 0, but len from 1
+	state.directionIndex = int32(len(p.cof.Priority) - 1)
+}
+
+func (p *COFWidget) deleteCurrentDirection() {
+	stateID := fmt.Sprintf("COFWidget_%s", p.id)
+	s := giu.Context.GetState(stateID)
+
+	state := s.(*COFState)
+	index := state.COFViewerState.directionIndex
+
+	p.cof.NumberOfDirections--
+
+	newPriority := make([][][]d2enum.CompositeType, 0)
+
+	for n, i := range p.cof.Priority {
+		if int32(n) != index {
+			newPriority = append(newPriority, i)
+		}
+	}
+
+	p.cof.Priority = newPriority
+}
+
+// CreateNewLayer starts add-cof-layer dialog
+func (p *COFWidget) CreateNewLayer() {
```

Comment:
we can probably break up this file further in another PR, but this looks good for now